lib/server/monalisa-synchronization/MonALISASynchronizer.js

```diff
@@ -0,0 +1,107 @@
+ *
+ * This software is distributed under the terms of the GNU General Public
+ * License v3 (GPL Version 3), copied verbatim in the file "COPYING".
+ *
+ * In applying this license CERN does not waive the privileges and immunities
+ * granted to it by virtue of its status as an Intergovernmental Organization
+ * or submit itself to any jurisdiction.
+ */
+const { Log } = require('@aliceo2/web-ui');
+const { setupMonALISAInterface } = require('./MonALISAInterface.js');
+const { dataSource } = require('../../database/DataSource.js');
+const { repositories: { DataPassRepository, RunRepository } } = require('../../database/index.js');
+const { extractLhcPeriod } = require('../utilities/extractLhcPeriod.js');
+const { getOrCreateLhcPeriod } = require('../services/lhcPeriod/getOrCreateLhcPeriod.js');
+
+/**
+ * Manger of synchronization with MonALISA
+ */
+class MonALISASynchronizer {
+    /**
+     * Constructor
+     */
+    constructor() {
+        this.logger = new Log(MonALISASynchronizer.name);
+        this.monALISAInterface = null;
+    }
+
+    /**
+     * Synchronize with data from MonALISA
+     * Update Data Passes data in local DB
+     * @return {Promise<void>} promise
+     */
+    async synchronizeDataPassesFromMonALISA() {
+        if (!this.monALISAInterface) {
+            this.monALISAInterface = await setupMonALISAInterface();
+        }
+
+        try {
+            this.lastRunNumbers = await this._getLastRunNumbersInformation();
+            const dataPasses = (await this.monALISAInterface.getDataPasses()).filter((dataPass) => this._doesDataPassNeedUpdate(dataPass));
+            // eslint-disable-next-line require-jsdoc
+            const transactionablePipeline = async (dataPass) => {
+                const lhcPeriodDB = await getOrCreateLhcPeriod({ name: extractLhcPeriod(dataPass.name).name });
+                await DataPassRepository.upsert({ lhcPeriodId: lhcPeriodDB.id, ...dataPass });
+                const dataPassDB = await DataPassRepository.findOne({ where: { name: dataPass.name } });
+                const { runNumbers } = await this.monALISAInterface.getDataPassDetails(dataPassDB.description);
+                const runs = await RunRepository.findAll(dataSource.createQueryBuilder().where('runNumber').oneOf(...runNumbers));
+                await dataPassDB.addRuns(runs, { ignoreDuplicates: true });
+            };
+
+            for (const dataPass of dataPasses) {
+                await dataSource.transaction(() => transactionablePipeline(dataPass));
+            }
+            this.logger.info('Synchronization completed');
+        } catch (error) {
+            this.logger.error(error.message);
+        } finally {
+            this.lastRunNumbers = null;
+        }
+    }
+
+    /**
+     * Check wether data pass needs to be updated
+     * @param {DataPass} dataPass data pass
+     * @return {boolean} true if it is needed to update data pass
+     */
+    _doesDataPassNeedUpdate(dataPass) {
```

Comment:
should take as parameter the last data pass run number

Reply:
applied